File: lib/domain/pest/interactors/pest_update_crop_associations_interactor.rb
# frozen_string_literal: true

module Domain
  module Pest
    module Interactors
      # Pest に紐づく crop 関連を差分同期する（認可済み crop_ids のみ渡すこと）。
      class PestUpdateCropAssociationsInteractor
        def initialize(crop_pest_gateway:)
          @sync = Services::CropPestAssociationSync.new(crop_pest_gateway: crop_pest_gateway)
        end

        # @param crop_ids [Array<Integer>]
        # @return [Domain::Pest::Dtos::PestCropAssociationSyncResult]
        def call(pest_id:, crop_ids:)
          @sync.replace_all(pest_id: pest_id, crop_ids: crop_ids)
        end
      end
    end
  end
end

File: crates/agrr-domain/src/pest/interactors/pest_update_crop_associations_interactor.rs
//! Ruby: `Domain::Pest::Interactors::PestUpdateCropAssociationsInteractor`

use crate::pest::dtos::PestCropAssociationSyncResult;
use crate::pest::gateways::CropPestGateway;
use crate::pest::services::CropPestAssociationSync;

pub struct PestUpdateCropAssociationsInteractor<'a, G> {
    sync: CropPestAssociationSync<'a, G>,
}

impl<'a, G: CropPestGateway> PestUpdateCropAssociationsInteractor<'a, G> {
    pub fn new(crop_pest_gateway: &'a G) -> Self {
        Self {
            sync: CropPestAssociationSync::new(crop_pest_gateway),
        }
    }

    pub fn call(
        &self,
        pest_id: i64,
        crop_ids: &[i64],
    ) -> Result<PestCropAssociationSyncResult, Box<dyn std::error::Error + Send + Sync>> {
        self.sync.replace_all(pest_id, crop_ids)
    }
}

#[cfg(test)]
mod interactors_pest_update_crop_associations_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/pest/interactors_pest_update_crop_associations_interactor_test.rs"));
}
